libraries/chain/controller.cpp

```diff
@@ -1902,9 +1923,12 @@ struct controller_impl {
 
          auto& ab = pending->_block_stage.get<assembled_block>();
 
-         // this implicitly asserts that all header fields (less the signature) are identical
-         EOS_ASSERT( producer_block_id == ab._id, block_validate_exception, "Block ID does not match",
-                     ("producer_block_id",producer_block_id)("validator_block_id",ab._id) );
+         if( producer_block_id != ab._id ) {
+            report_block_header_diff( *b, *ab._unsigned_block );
```

Comment:
This will end up putting up some fairly confusing log statements.  It would be good to add some context here of what "Block ID does not match" even if it was just as simple as "assembled block does not match received block do to the above block_header fields not matching".  Also, I think it would be good here to have a comment indicating what the assembled block is (i.e. why do we assemble the block from the received block and what is different about it - philosophy wise, not content wise, which report_block_header_diff will give you, like is it just recalculating the header), or at least some kind of comment indicating what this means for the received block not to match the assembled block.